daphne/src/messages/mod.rs

```diff
@@ -1378,24 +1402,30 @@ pub fn decode_base64url_vec<T: AsRef<[u8]>>(input: T) -> Option<Vec<u8>> {
     URL_SAFE_NO_PAD.decode(input).ok()
 }
 
-// Cribbed from `decode_u16_items()` from libprio.
-fn encode_u16_prefixed(
-    version: DapVersion,
-    bytes: &mut Vec<u8>,
-    e: impl Fn(DapVersion, &mut Vec<u8>) -> Result<(), CodecError>,
-) -> Result<(), CodecError> {
-    // Reserve space for the length prefix.
-    let len_offset = bytes.len();
-    0_u16.encode(bytes)?;
-
-    e(version, bytes)?;
-    let len_bytes = std::mem::size_of::<u16>();
-    let len = bytes.len() - len_offset - len_bytes;
-    bytes[len_offset..len_offset + len_bytes]
-        .copy_from_slice(&u16::to_be_bytes(len.try_into().unwrap()));
-    Ok(())
+macro_rules! make_encode_len_prefixed {
+    ($type:ident, $name:ident) => {
+        pub(crate) fn $name(
+            version: DapVersion,
+            bytes: &mut Vec<u8>,
+            e: impl Fn(DapVersion, &mut Vec<u8>) -> Result<(), CodecError>,
+        ) -> Result<(), CodecError> {
+            // Reserve space for the length prefix.
+            let len_offset = bytes.len();
+            (0 as $type).encode(bytes)?;
+
+            e(version, bytes)?;
+            let len_bytes = std::mem::size_of::<$type>();
+            let len = bytes.len() - len_offset - len_bytes;
+            bytes[len_offset..len_offset + len_bytes]
+                .copy_from_slice(&$type::to_be_bytes(len.try_into().unwrap()));
+            Ok(())
+        }
+    };
 }
 
+make_encode_len_prefixed!(u16, encode_u16_prefixed);
+make_encode_len_prefixed!(u32, encode_u32_prefixed);
+
```

Comment:
This can probably be implemented as a generic function
```rust
where
  T: Default + Encode
  ```
and instead of using std::mem::size_of you can do
```rust
let len_bytes = bytes.len() - len_offset;
```

Reply:
thinking back on it you can't be generic over `to_be_bytes` but at the same time, should you be using that at all? Shouldn't you be using `len.encode`? I understand that the `encode` api doesn't let you encode in the middle, but at the same time, doesn't this code only work for as long as the encode implementation of u16 and u32 doesn't change?

Reply:
That might work, I'll give it a shot.

Reply:
yeah, I agree with the downsides you've listed, they might not be worth it. Keep the macro for now